src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs

```diff
@@ -1345,10 +1421,52 @@ private void AssignPatternVariables(BoundPattern pattern)
         public override BoundNode VisitBlock(BoundBlock node)
         {
             DeclareVariables(node.Locals);
-            var result = base.VisitBlock(node);
+
+            VisitStatementsWithLocalFunctions(node);
+
             ReportUnusedVariables(node.Locals);
             ReportUnusedVariables(node.LocalFunctions);
-            return result;
+
+            return null;
+        }
+
+        protected void VisitStatementsWithLocalFunctions(BoundBlock block)
+        {
+            // Visit the statements in two phases:
+            //   1. Local function declarations
+            //   2. Everything else
+            //
+            // The idea behind visiting local functions first is
+            // that we may be able to gather the captured variables
+            // they read and write ahead of time in a single pass, so
+            // when they are used by other statements in the block we
+            // won't have to recompute the set by doing multiple passes.
+            //
+            // If the local functions contain forward calls to other local
+            // functions then we may have to do another pass regardless,
+            // but hopefully that will be an uncommon case in real-world code.
+
+            // First phase
+            if (!block.LocalFunctions.IsDefaultOrEmpty)
+            {
+                foreach (var stmt in block.Statements)
+                {
+                    if (stmt.Kind == BoundKind.LocalFunctionStatement)
+                    {
+                        VisitLocalFunctionStatement(
+                            (BoundLocalFunctionStatement)stmt);
+                    }
+                }
+            }
```

Comment:
If you're not going to process the statements fully by replaying the local functions back, you might as well place a loop around the first phase (while the data for any of *these* local functions has changed). Because otherwise any work you do on the regular statements, below, is just wasted.

The same logic applies to any block in a nested local function too.